includes/tracks/class-wc-tracks.php

```diff
@@ -0,0 +1,155 @@
+			return array(
+				'_ut' => 'wpcom:user_id',
+				'_ui' => $wpcom_id,
+			);
+		}
+
+		// User is connected, but no meta is set yet.  Use WPCOM ID and set meta.
+		if ( $has_jetpack && Jetpack::is_user_connected( $user_id ) ) {
+			$wpcom_user_data = Jetpack::get_connected_user_data( $user_id );
+			add_user_meta( $user_id, 'jetpack_tracks_wpcom_id', $wpcom_user_data['ID'], true );
+
+			return array(
+				'_ut' => 'wpcom:user_id',
+				'_ui' => $wpcom_user_data['ID'],
+			);
+		}
+
+		// User isn't linked at all.  Fall back to anonymous ID.
+		$anon_id = get_user_meta( $user_id, 'jetpack_tracks_anon_id', true );
+		if ( ! $anon_id ) {
+			$anon_id = WC_Tracks_Client::get_anon_id();
+			add_user_meta( $user_id, 'jetpack_tracks_anon_id', $anon_id, false );
+		}
+
+		if ( ! isset( $_COOKIE['tk_ai'] ) && ! headers_sent() ) {
+			setcookie( 'tk_ai', $anon_id );
+		}
+
+		return array(
+			'_ut' => 'anon',
+			'_ui' => $anon_id,
+		);
+
+	}
+
+	/**
+	 * Gather blog related properties.
+	 *
+	 * @param int $user_id User id.
+	 * @return array Blog details.
+	 */
+	public static function get_blog_details( $user_id ) {
+		return array(
+			// @TODO: Add revenue/product info and url similar to wc-tracker
+			'url'       => get_option( 'siteurl' ),
+			'blog_lang' => get_user_locale( $user_id ),
+			'blog_id'   => ( class_exists( 'Jetpack' ) && Jetpack_Options::get_option( 'id' ) ) || null,
+		);
+	}
+
+	/**
+	 * Gather details from the request to the server.
+	 *
+	 * @return array Server details.
+	 */
+	public static function get_server_details() {
+		$data = array();
+
+		$data['_via_ua'] = isset( $_SERVER['HTTP_USER_AGENT'] ) ? $_SERVER['HTTP_USER_AGENT'] : '';
+		$data['_via_ip'] = isset( $_SERVER['REMOTE_ADDR'] ) ? $_SERVER['REMOTE_ADDR'] : '';
+		$data['_lg']     = isset( $_SERVER['HTTP_ACCEPT_LANGUAGE'] ) ? $_SERVER['HTTP_ACCEPT_LANGUAGE'] : '';
+		$data['_dr']     = isset( $_SERVER['HTTP_REFERER'] ) ? $_SERVER['HTTP_REFERER'] : '';
+
+		$uri         = isset( $_SERVER['REQUEST_URI'] ) ? $_SERVER['REQUEST_URI'] : '';
+		$host        = isset( $_SERVER['HTTP_HOST'] ) ? $_SERVER['HTTP_HOST'] : '';
+		$data['_dl'] = $_SERVER['REQUEST_SCHEME'] . '://' . $host . $uri;
+
+		return $data;
+	}
+
+	/**
+	 * Record an event in Tracks - this is the preferred way to record events from PHP.
+	 *
+	 * @param string $event_name The name of the event.
+	 * @param array  $properties Custom properties to send with the event.
+	 * @return bool|WP_Error true for success or WP_Error if the event pixel could not be fired.
+	 */
+	public static function record_event( $event_name, $properties = array() ) {
+		$user = wp_get_current_user();
+
+		// We don't want to track user events during unit tests/CI runs.
+		if ( $user instanceof WP_User && 'wptests_capabilities' === $user->cap_key ) {
+			return false;
+		}
+
+		/**
+		 * Don't track users who haven't opted-in to tracking or if a filter
+		 * has been applied to turn it off.
+		 */
+		if (
```

Comment:
Why not check this earlier/first before validating user etc?

Reply:
Absolutely. In https://github.com/woocommerce/woocommerce/pull/22746 I've added a check so none of the actions are even added if a user is opted out. 

I think it makes sense to leave the check here as well in case this class is used elsewhere.

Reply:
Could be done earlier in this method also since it disables tracking.